### PR TITLE
[ONNX] Extra support for bernoulli export

### LINKED ISSUE #88299: [ONNX] support extra parameters for Bernoulli when exporting
### 🐛 Describe the bug

Tried exporting a torch model with code like `x.bernoulli_(p)` but failed.
Error is also thrown when exporting `x.bernoulli(p)` or `x.bernoulli_()`. Only `x.bernoulli()` is exported successfully.

Minimal repro case:
```py
import torch

class Net(torch.nn.Module):
    def forward(self, x):
        # return x.bernoulli(0.2)  # case #1
        # return x.bernoulli_()    # case #2
        # return x.bernoulli_(0.2) # case #3
        return x.bernoulli()       # case #4

model = Net()
x = torch.rand(4, 5)
torch.onnx.export(model, x, 'bernoulli_export.onnx', opset_version=15)
```
where
* cases 1&2&3: fail with error `RuntimeError: Unsupported: ONNX export of operator Bernoulli, out parameter is not supported for bernoulli. Please feel free to request support or submit a pull request on PyTorch GitHub.`
* case 4: successfully exported.

Hope cases 1-3 would also be supported for exporting, thanks.

### Versions

```text
Collecting environment information...
PyTorch version: 1.12.1
Is debug build: False
CUDA used to build PyTorch: 11.6
ROCM used to build PyTorch: N/A

OS: Ubuntu 20.04.5 LTS (x86_64)
GCC version: (Ubuntu 9.4.0-1ubuntu1~20.04.1) 9.4.0
Clang version: Could not collect
CMake version: version 3.23.0
Libc version: glibc-2.31

Python version: 3.8.13 (default, Mar 28 2022, 11:38:47)  [GCC 7.5.0] (64-bit runtime)
Python platform: Linux-5.4.0-1063-azure-x86_64-with-glibc2.17
Is CUDA available: True
CUDA runtime version: 11.6.124
CUDA_MODULE_LOADING set to: 
GPU models and configuration: 
GPU 0: NVIDIA A100-SXM4-40GB
GPU 1: NVIDIA A100-SXM4-40GB
GPU 2: NVIDIA A100-SXM4-40GB
GPU 3: NVIDIA A100-SXM4-40GB
GPU 4: NVIDIA A100-SXM4-40GB
GPU 5: NVIDIA A100-SXM4-40GB
GPU 6: NVIDIA A100-SXM4-40GB
GPU 7: NVIDIA A100-SXM4-40GB

Nvidia driver version: 470.82.01
cuDNN version: Probably one of the following:
/usr/lib/x86_64-linux-gnu/libcudnn.so.8.4.0
/usr/lib/x86_64-linux-gnu/libcudnn_adv_infer.so.8.4.0
/usr/lib/x86_64-linux-gnu/libcudnn_adv_train.so.8.4.0
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_infer.so.8.4.0
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_train.so.8.4.0
/usr/lib/x86_64-linux-gnu/libcudnn_ops_infer.so.8.4.0
/usr/lib/x86_64-linux-gnu/libcudnn_ops_train.so.8.4.0
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

Versions of relevant libraries:
[pip3] numpy==1.22.2
[pip3] pytorch-lightning==1.6.3
[pip3] torch==1.12.1
[pip3] torch-nebula==0.15.1
[pip3] torch-ort==1.12.0
[pip3] torchlars==0.1.2
[pip3] torchmetrics==0.7.1
[pip3] torchvision==0.13.1+cu116
[conda] magma-cuda116             2.6.1                         1    pytorch
[conda] mkl                       2021.4.0                 pypi_0    pypi
[conda] mkl-include               2021.4.0                 pypi_0    pypi
[conda] numpy                     1.22.2                   pypi_0    pypi
[conda] pytorch-lightning         1.6.3                    pypi_0    pypi
[conda] torch                     1.12.1                   pypi_0    pypi
[conda] torch-nebula              0.15.1                   pypi_0    pypi
[conda] torch-ort                 1.12.0                   pypi_0    pypi
[conda] torchlars                 0.1.2                    pypi_0    pypi
[conda] torchmetrics              0.7.1                    pypi_0    pypi
[conda] torchvision               0.13.1+cu116             pypi_0    pypi
```

### PR DESCRIPTION
**Pull request by mindest**
* add opset 15 support for `bernoulli`.
* add extra export options for different `bernoulli` cases: `x.bernoulli(p)` where `p` is a tensor or float.

Fixes #88299

### CODE REVIEW

File: torch/onnx/symbolic_opset15.py [hunk 1]
```diff
@@ -54,6 +54,22 @@ def aten__isnot_(g: jit_utils.GraphContext, self, other):
...
+        symbolic_helper._unimplemented(
+            "Bernoulli", "out parameter is not supported for bernoulli", input
+        )
+    if generator is not None and not symbolic_helper._is_none(generator):
+        symbolic_helper._unimplemented(
+            "Bernoulli", "generator is not supported for bernoulli", input
+        )
+    if tensor_or_prob is None or symbolic_helper._is_none(tensor_or_prob):
+        return g.op("Bernoulli", input)
+    return opset9.bernoulli(g, input, tensor_or_prob, generator, out)
```
Comment (justinchuby) [inline torch/onnx/symbolic_opset15.py:70]: Thanks for the contribution. For my understanding, is there a reason why there is a need to override the opset9 implementation? Does the behavior change for opset 15?

**Comment by mindest** [comment]
Hi @AllenTiTaiWang @justinchuby, if I understand correctly, in the existing test it does not check correctness of bernoulli right? It is equivalent to tensor `x` multiplied by a scalar. So what's the main purpose of this test?
https://github.com/pytorch/pytorch/blob/53ca5ad347451f3087dedc8df5c1a34663812a6b/test/onnx/test_pytorch_onnx_onnxruntime.py#L2638

**Comment by AllenTiTaiWang** [comment]
> Hi @AllenTiTaiWang @justinchuby, if I understand correctly, in the existing test it does not check correctness of bernoulli right? It is equivalent to tensor `x` multiplied by a scalar. So what's the main purpose of this test?
> 
> https://github.com/pytorch/pytorch/blob/53ca5ad347451f3087dedc8df5c1a34663812a6b/test/onnx/test_pytorch_onnx_onnxruntime.py#L2638

I am not sure. I guess the author got this example somewhere else. Could you add a new test about bernoulli? Let's leave this there (You can change its name though).

**Comment by justinchuby** [comment]
bernoulli is a random op so I think it's harder to test numerically? Maybe just test for the statistics, or simply verify a valid graph?

File: torch/onnx/symbolic_opset9.py [hunk 2]
```diff
@@ -4942,8 +4942,8 @@ def rrelu(g: jit_utils.GraphContext, input, lower, upper, training, generator):
 
 @_onnx_symbolic("aten::bernoulli")
 @_beartype.beartype
-def bernoulli(g: jit_utils.GraphContext, input, generator=None, out=None):
-    if out is not None:
+def bernoulli(g: jit_utils.GraphContext, input, tensor_or_prob=None, generator=None, out=None):
```
Comment (justinchuby) [inline torch/onnx/symbolic_opset9.py:4945]: I don't need a `tensor_or_prob` argument in the native_functions.yaml definition. Is this new?

Comment (justinchuby) [reply torch/onnx/symbolic_opset9.py:4945]: I see its called `p`. I would just call it that?

(Seen)

```
- func: bernoulli_.Tensor(Tensor(a!) self, Tensor p, *, Generator? generator=None) -> Tensor(a!)
  device_check: NoCheck   # TensorIterator
  variants: method
  tags: nondeterministic_seeded
  dispatch:
    CPU, CUDA: bernoulli_
    MPS: bernoulli_mps_
  autogen: bernoulli.Tensor, bernoulli.Tensor_out

- func: bernoulli_.float(Tensor(a!) self, float p=0.5, *, Generator? generator=None) -> Tensor(a!)
  device_check: NoCheck   # TensorIterator
  variants: method
  tags: nondeterministic_seeded
  dispatch:
    CPU, CUDA: bernoulli_
    MPS: bernoulli_mps_
  autogen: bernoulli.float_out
```

Comment (mindest) [reply torch/onnx/symbolic_opset9.py:4945]: Sure, I will change it to `p` and the existing `p` to another name.

**Comment by mindest** [comment]
> bernoulli is a random op so I think it's harder to test numerically? Maybe just test for the statistics, or simply verify a valid graph?

Yes I understand it is hard for these tests to verify random values. I've just added some tests similar to the existing one, which in fact do not check correctness of bernoulli or verify the graph. If needed pls let me know.

**Comment by AllenTiTaiWang** [comment]
> > bernoulli is a random op so I think it's harder to test numerically? Maybe just test for the statistics, or simply verify a valid graph?
> 
> Yes I understand it is hard for these tests to verify random values. I've just added some tests similar to the existing one, which in fact do not check correctness of bernoulli or verify the graph. If needed pls let me know.

I guess it's fine for the Op, because in converter, we simply replace the node, and no other processing for upstream parameters or input, @justinchuby what do you think?

**Comment by justinchuby** [comment]
SGTM. I wonder why 

        self.run_test(Bernoulli_float(), x)
        self.run_test(Bernoulli_tensor(), x)

would pass though?

**Comment by AllenTiTaiWang** [comment]
@BowenBao PTAL

**Review by BowenBao — APPROVED** [review]

**Comment by AllenTiTaiWang** [comment]
@mindest your PR is ready to go, check out https://github.com/pytorch/pytorch/wiki/Bot-commands to merge it when you are ready.

**Pull request closed without merging**